File: src/store/mouse.ts
import { create } from "zustand";
import { invoke } from "@tauri-apps/api/core";
import { listen } from "@tauri-apps/api/event";

// Mirrors the Rust structs in src-tauri/src/mouse/.
export interface MouseStatus {
  present: boolean;
  model: string | null;
  wireless: boolean;
  battery_percent: number | null;
  charging: boolean;
}

export interface MouseConfig {
  dpi_presets: number[];
  dpi_selected: number;
  polling_hz: number;
  zones: [number, number, number][];
  rainbow: boolean;
  reactive: [number, number, number] | null;
  sleep_minutes: number;
  dim_seconds: number;
}

interface MouseSnapshot {
  connected: boolean;
  status: MouseStatus;
  config: MouseConfig;
  dpi_steps: number[];
  polling_rates: number[];
}

const emptyStatus: MouseStatus = {
  present: false,
  model: null,
  wireless: false,
  battery_percent: null,
  charging: false,
};

const defaultConfig: MouseConfig = {
  dpi_presets: [400, 800, 1200, 2400, 3200],
  dpi_selected: 0,
  polling_hz: 1000,
  zones: [
    [255, 0, 0],
    [0, 255, 0],
    [0, 0, 255],
  ],
  rainbow: true,
  reactive: null,
  sleep_minutes: 5,
  dim_seconds: 30,
};

/** Zone labels in device order. */
export const ZONE_LABELS = ["Top", "Middle", "Bottom"] as const;

interface MouseState {
  connected: boolean;
  status: MouseStatus;
  config: MouseConfig;
  dpiSteps: number[];
  pollingRates: number[];
  error: string | null;
  _initialized: boolean;

  init: () => Promise<void>;
  refresh: () => Promise<void>;
  setDpi: (presets: number[], selected: number) => Promise<void>;
  setPolling: (hz: number) => Promise<void>;
  setZoneColor: (zone: number, rgb: [number, number, number]) => Promise<void>;
  setRainbow: () => Promise<void>;
  setReactive: (rgb: [number, number, number] | null) => Promise<void>;
  setSleep: (minutes: number) => Promise<void>;
  setDim: (seconds: number) => Promise<void>;
}

export const useMouse = create<MouseState>((set, get) => {
  // Device writes are slow (the mouse needs ~50 ms between commands), so a
  // colour-picker drag must not fire one invoke per pixel.
  const pending = new Map<string, ReturnType<typeof setTimeout>>();
  const debounced = (key: string, fn: () => Promise<unknown>, ms = 200) => {
    const prev = pending.get(key);
    if (prev) clearTimeout(prev);
    pending.set(
      key,
      setTimeout(() => {
        pending.delete(key);
        void fn().catch((e) => set({ error: String(e) }));
      }, ms),
    );
  };

  return {
    connected: false,
    status: emptyStatus,
    config: defaultConfig,
    dpiSteps: [],
    pollingRates: [125, 250, 500, 1000],
    error: null,
    _initialized: false,

    refresh: async () => {
      const snap = await invoke<MouseSnapshot>("get_mouse_status");
      set({
        connected: snap.connected,
        status: snap.status,
        config: snap.config,
        dpiSteps: snap.dpi_steps,
        pollingRates: snap.polling_rates,
      });
    },

    init: async () => {
      if (get()._initialized) return;
      set({ _initialized: true });
      try {
        await get().refresh();
      } catch (e) {
        console.error("mouse init:", e);
      }
      void listen<MouseStatus>("mouse-status", (e) => set({ status: e.payload }));
      void listen<boolean>("mouse-presence", (e) => {
        set((s) => ({
          connected: e.payload,
          status: e.payload ? s.status : emptyStatus,
        }));
        if (e.payload) void get().refresh();
      });
    },

    setDpi: async (presets, selected) => {
      set((s) => ({
        config: { ...s.config, dpi_presets: presets, dpi_selected: selected },
      }));
      debounced("dpi", () => invoke("mouse_set_dpi", { presets, selected }));
    },
    setPolling: async (hz) => {
      set((s) => ({ config: { ...s.config, polling_hz: hz } }));
      await invoke("mouse_set_polling", { hz });
    },
    setZoneColor: async (zone, rgb) => {
      set((s) => {
        const zones = s.config.zones.map((z, i) => (i === zone ? rgb : z));
        return { config: { ...s.config, zones, rainbow: false } };
      });
      debounced(`zone:${zone}`, () =>
        invoke("mouse_set_zone_color", { zone, r: rgb[0], g: rgb[1], b: rgb[2] }),
      );
    },
    setRainbow: async () => {
      set((s) => ({ config: { ...s.config, rainbow: true } }));
      await invoke("mouse_set_rainbow");
    },
    setReactive: async (rgb) => {
      set((s) => ({ config: { ...s.config, reactive: rgb } }));
      await invoke("mouse_set_reactive", { rgb });
    },
    setSleep: async (minutes) => {
      set((s) => ({ config: { ...s.config, sleep_minutes: minutes } }));
      debounced("sleep", () => invoke("mouse_set_sleep", { minutes }));
    },
    setDim: async (seconds) => {
      set((s) => ({ config: { ...s.config, dim_seconds: seconds } }));
      debounced("dim", () => invoke("mouse_set_dim", { seconds }));
    },
  };
});

File: src-tauri/src/commands/mouse.rs
//! Tauri commands for the SteelSeries mouse (Aerox 9 Wireless).

use serde::Serialize;
use tauri::State;

use crate::mouse::manager::MouseStatus;
use crate::mouse::protocol::{self, MouseConfig};
use crate::state::AppState;

#[derive(Serialize)]
pub struct MouseSnapshot {
    pub connected: bool,
    pub status: MouseStatus,
    pub config: MouseConfig,
    /// Selectable DPI steps the sensor actually supports.
    pub dpi_steps: Vec<u16>,
    pub polling_rates: Vec<u16>,
}

#[tauri::command]
pub fn get_mouse_status(state: State<'_, AppState>) -> MouseSnapshot {
    MouseSnapshot {
        connected: state.mouse.is_connected(),
        status: state.mouse.status(),
        config: state.mouse.config(),
        dpi_steps: protocol::dpi_steps(),
        polling_rates: protocol::POLLING_RATES.to_vec(),
    }
}

/// Set the DPI preset list and which entry is active, then persist.
#[tauri::command]
pub fn mouse_set_dpi(
    state: State<'_, AppState>,
    presets: Vec<u16>,
    selected: u8,
) -> Result<(), String> {
    state
        .mouse
        .send_with_pid(|pid| protocol::set_dpi(pid, &presets, selected))?;
    let mut cfg = state.mouse.config();
    cfg.dpi_presets = presets;
    cfg.dpi_selected = selected;
    state.mouse.set_config(cfg);
    state.mouse.save()
}

#[tauri::command]
pub fn mouse_set_polling(state: State<'_, AppState>, hz: u16) -> Result<(), String> {
    state.mouse.send_with_pid(|pid| protocol::set_polling(pid, hz))?;
    let mut cfg = state.mouse.config();
    cfg.polling_hz = hz;
    state.mouse.set_config(cfg);
    state.mouse.save()
}

/// Set one lighting zone's colour (0 = top, 1 = middle, 2 = bottom).
#[tauri::command]
pub fn mouse_set_zone_color(
    state: State<'_, AppState>,
    zone: u8,
    r: u8,
    g: u8,
    b: u8,
) -> Result<(), String> {
    state
        .mouse
        .send_with_pid(|pid| protocol::set_zone_color(pid, zone, r, g, b))?;
    let mut cfg = state.mouse.config();
    // Setting a solid colour implicitly turns the rainbow effect off.
    cfg.rainbow = false;
    if let Some(slot) = cfg.zones.get_mut(zone as usize) {
        *slot = [r, g, b];
    }
    state.mouse.set_config(cfg);
    state.mouse.save()
}

/// Enable the rainbow effect across all zones.
#[tauri::command]
pub fn mouse_set_rainbow(state: State<'_, AppState>) -> Result<(), String> {
    state.mouse.send_with_pid(protocol::set_rainbow)?;
    let mut cfg = state.mouse.config();
    cfg.rainbow = true;
    state.mouse.set_config(cfg);
    state.mouse.save()
}

/// Reactive click-flash colour; pass `null` to disable.
#[tauri::command]
pub fn mouse_set_reactive(
    state: State<'_, AppState>,
    rgb: Option<[u8; 3]>,
) -> Result<(), String> {
    let tuple = rgb.map(|c| (c[0], c[1], c[2]));
    state
        .mouse
        .send_with_pid(|pid| protocol::set_reactive(pid, tuple))?;
    let mut cfg = state.mouse.config();
    cfg.reactive = rgb;
    state.mouse.set_config(cfg);
    state.mouse.save()
}

/// Sleep timeout in minutes (0 = never, max 20).
#[tauri::command]
pub fn mouse_set_sleep(state: State<'_, AppState>, minutes: u16) -> Result<(), String> {
    state
        .mouse
        .send_with_pid(|pid| protocol::set_sleep_minutes(pid, minutes))?;
    let mut cfg = state.mouse.config();
    cfg.sleep_minutes = minutes;
    state.mouse.set_config(cfg);
    state.mouse.save()
}

/// Which lighting effects come back after the mouse power-cycles.
#[tauri::command]
pub fn mouse_set_startup_lighting(
    state: State<'_, AppState>,
    rainbow: bool,
    reactive: bool,
) -> Result<(), String> {
    state
        .mouse
        .send_with_pid(|pid| protocol::set_startup_lighting(pid, rainbow, reactive))?;
    state.mouse.save()
}

/// Light-dim timeout in seconds (0 = never, max 1200).
#[tauri::command]
pub fn mouse_set_dim(state: State<'_, AppState>, seconds: u16) -> Result<(), String> {
    state
        .mouse
        .send_with_pid(|pid| protocol::set_dim_seconds(pid, seconds))?;
    let mut cfg = state.mouse.config();
    cfg.dim_seconds = seconds;
    state.mouse.set_config(cfg);
    state.mouse.save()
}
